include/nbl/builtin/hlsl/bxdf/ndf.hlsl

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BXDF_NDF_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BXDF_NDF_INCLUDED_
+
+#include "nbl/builtin/hlsl/limits.hlsl"
+#include "nbl/builtin/hlsl/bxdf/common.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace bxdf
+{
+namespace ndf
+{
+
+// common
+namespace impl
+{
+template<typename T, bool ggx = false>
+struct microfacet_to_light_measure_transform
+{
+    static T __call(T NDFcos, T absNdotV, bool transmitted, T VdotH, T LdotH, T VdotHLdotH, T orientedEta)
+    {
+        T denominator = absNdotV;
+        if (transmitted)
+        {
+            const T VdotH_etaLdotH = (VdotH+orientedEta*LdotH);
+            // VdotHLdotH is negative under transmission, so thats denominator is negative
+            denominator *= -VdotH_etaLdotH * VdotH_etaLdotH;
+        }
+        return NDFcos * (transmitted ? VdotHLdotH : 0.25) / denominator;
+    }
+
+    static T __call(T NDFcos, T maxNdotV)
+    {
+        return 0.25 * NDFcos / maxNdotV;
+    }
+
+};
+
+template<typename T>
+struct microfacet_to_light_measure_transform<T,true>
+{
+    static T __call(T NDFcos_already_in_reflective_dL_measure, T absNdotL, bool transmitted, T VdotH, T LdotH, T VdotHLdotH, T orientedEta)
+    {
+        T denominator = absNdotL;
+        if (transmitted)
+        {
+            const T VdotH_etaLdotH = (VdotH+orientedEta*LdotH);
+            // VdotHLdotH is negative under transmission, so thats denominator is negative
+            denominator *= -VdotH_etaLdotH * VdotH_etaLdotH;
+        }
+        return NDFcos_already_in_reflective_dL_measure * (transmitted ? VdotHLdotH : 0.25) / denominator;
+    }
+
+    static T __call(T NDFcos_already_in_reflective_dL_measure, T maxNdotL)
+    {
+        return NDFcos_already_in_reflective_dL_measure * maxNdotL;
+    }
+};
+}
+
+template<typename T, bool ggx NBL_FUNC_REQUIRES(is_scalar_v<T>)
+T microfacet_to_light_measure_transform(T NDFcos, T absNdotV, bool transmitted, T VdotH, T LdotH, T VdotHLdotH, T orientedEta)
+{
+    return impl::microfacet_to_light_measure_transform<T,ggx>::__call(NDFcos, absNdotV, transmitted, VdotH, LdotH, VdotHLdotH, orientedEta);
+}
+
+template<typename T, bool ggx NBL_FUNC_REQUIRES(is_scalar_v<T>)
+T microfacet_to_light_measure_transform(T NDFcos, T maxNdotV)
+{
+    return impl::microfacet_to_light_measure_transform<T,ggx>::__call(NDFcos, maxNdotV);
+}
```

Comment:
just have the thing as a struct with partial spec, no need to have it a function.

Also same thing with the arguments, turn into members

Reply:
Also imho it would be best to split this into 6 cases:
- GGX vs non GGX
- BxDF, refract only, reflect only (brdf doesn't transmit/refract, so no eta needed)

you can template on an enum
```cpp
REFLECT_BIT = 0b01,
REFRACT_BIT = 0b10,
REFLECT_REFRACT_BIT = 0b11
```